### PR TITLE
[WIP][SPARK-31363][SQL] Add DataSourceRegisterV2

### PR DESCRIPTION
**Pull request by PerilousApricot**
### What changes were proposed in this pull request?
Add DataSourceRegisterV2, an alternative method for DataSourceV2 plugins to register themselves. The motivation is to add a layer of indirection between the data source registration and data source provider, so plugin developers have a chance choose at runtime which implementation Spark should use.  Discussion at:

http://apache-spark-developers-list.1001551.n3.nabble.com/DSv2-amp-DataSourceRegister-td29187.html

### Why are the changes needed?
As the DSv2 API evolves, some breaking changes are occasionally made to the API. It's possible to split a plugin into a "common" part and multiple version-specific parts and this works good to have a single artifact for users. The one part that can't be currently worked around is the DataSourceRegister trait. This is an issue because users cargo-cult configuration values, and choosing the wrong plugin version gives a particularly baroque error message that bubbles up through ServiceLoader.

### Does this PR introduce any user-facing change?
No, developer only.

### How was this patch tested?

Unit tests were added

### CODE REVIEW

**Comment by PerilousApricot** [comment]
I did a POC with Spark2.4 as well, which can be found at https://github.com/PerilousApricot/spark/tree/feature/registerv2-24

I compiled my datasource and both Spark patches and verified that the patch correctly loads the right version of the plugin. You can verify it yourself with the following jar http://mirror.accre.vanderbilt.edu/spark/laurelin-both.jar and input file https://github.com/spark-root/laurelin/raw/master/testdata/stdvector.root

```
scala> val df = spark.read.format("root").option("tree","tvec").load("stdvector.root")
val df = spark.read.format("root").option("tree","tvec").load("stdvector.root")
df: org.apache.spark.sql.DataFrame = [vpx: array<float>, vpy: array<float> ... 7 more fields]
scala> df.select("vpx").show()
+--------------------+
|                 vpx|
+--------------------+
|         [-2.409915]|
|[-0.39048654, 0.3...|
|[-0.0517636, 0.06...|
|[-0.4088529, 0.23...|
|[0.22863834, -0.3...|
|[1.5184512, -1.10...|
|[-1.2615219, 1.17...|
|[0.26021498, -0.3...|
|[-1.2066879, 0.91...|
|[0.54583037, -0.4...|
+--------------------+
```

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 1]
```diff
@@ -50,6 +50,40 @@ trait DataSourceRegister {
...
+/**
+ * Extended version of [[DataSourceRegister]] which allows a data source to
+ * select an implementing class at runtime.  This allows plugin developers to
+ * choose the correct implementation at runtime, depending on the version of
+ * Spark the user is using.
+ *
+ *  A new instance of this class will be instantiated each time a DDL call is made.
+ *
+ */
+trait DataSourceRegisterV2 {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:62]: I think this works for `TableProvider`, which is used in `DataFrameReader/Writer`. But `TableCatalog` is a different API and it uses a different way to plug into Spark, so this won't work.

cc @rdblue @brkyvz do you think this is valuable?

**Comment by PerilousApricot** [comment]
Hi, this is meant strictly as an improvement to the DataSourceRegister API,
which also isn't used by the catalog.
-- 
It's dark in this basement.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala (hunk 1 above)
Comment (PerilousApricot) [reply sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala:62]: This is meant strictly as an improvement to the DataSourceRegister API,
which also isn't used by the catalog.

**Comment by PerilousApricot** [comment]
@cloud-fan @rdblue @brkyvz Now that 3.0 has landed and we're in the situation where some universities have moved to 3.0 while others are @ 2.4, and there's a chance changes planned for 3.1 will change the DSv2 interfaces again, it would be really nice to add this interface to support these (globally) heterogeneous deployments. I can rebase the patches if that will make examining it similar

**Pull request closed without merging**